jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c

```diff
@@ -642,59 +677,66 @@ ecma_builtin_array_prototype_object_shift (ecma_object_t *obj_p, /**< array obje
   }
 
   /* 6. and 7. */
-  for (uint32_t k = 1; k < len && ecma_is_value_empty (ret_value); k++)
+  for (uint32_t k = 1; k < len; k++)
   {
-    /* 7.a */
-    ecma_string_t *from_str_p = ecma_new_ecma_string_from_uint32 (k);
-    /* 7.b */
-    ecma_string_t *to_str_p = ecma_new_ecma_string_from_uint32 (k - 1);
+    /* 7.a - 7.c */
+    ecma_value_t curr_value = ecma_op_object_find_by_index (obj_p, k);
 
-    /* 7.c */
-    ECMA_TRY_CATCH (curr_value, ecma_op_object_find (obj_p, from_str_p), ret_value);
+    if (ECMA_IS_VALUE_ERROR (curr_value))
+    {
+      ecma_free_value (first_value);
+      return curr_value;
+    }
+
+    /* 7.b */
+    uint32_t to = k - 1;
 
     if (ecma_is_value_found (curr_value))
     {
       /* 7.d.i, 7.d.ii */
-      ECMA_TRY_CATCH (put_value, ecma_op_object_put (obj_p, to_str_p, curr_value, true), ret_value);
+      ecma_value_t put_value = ecma_op_object_put_by_index (obj_p, to, curr_value, true);
```

Comment:
The "put" and "delete" parts are really similar, would it be possible to refactor it like this?
```c
ecma_value_t new_value;

if (ecma_value_is_found (curr_value))
{
    new_value = ecma_op_object_put_by_index (obj_p, to, curr_value, true);
    ecma_free_value (curr_value);
}
else
{
    new_value = ecma_op_object_delete_by_index (obj_p, to, true);
}

if (ECMA_IS_VALUE_ERROR (new_value))
{
    ecma_free_value (first_value);
    return new_value;
}
```

I'm okay having it in a followup PR.

Reply:
Good idea, let me do it in a follow up patch.

Reply:
@galpeter I've already updated the PR with this idea.